Packet++/src/SomeIpLayer.cpp

```diff
@@ -0,0 +1,324 @@
+#define LOG_MODULE PacketLogModuleSomeIpLayer
+
+#include "SomeIpLayer.h"
+#include "EndianPortable.h"
+#include "Packet.h"
+#include "PayloadLayer.h"
+#include <algorithm>
+#include <sstream>
+#include <string.h>
+
+namespace pcpp
+{
+
+void splitUint32Id(uint32_t uint32Id, uint16_t &uint16IdUpper, uint16_t &uint16IdLower)
+{
+	uint16IdLower = (uint32Id & 0x0000ffff);
+	uint16IdUpper = (uint32Id & 0xffff0000) >> 16;
+}
+
+std::unordered_set<uint16_t> SomeIpLayer::m_SomeIpPorts{};
+
+SomeIpLayer::SomeIpLayer(uint16_t serviceID, uint16_t methodID, uint16_t clientID, uint16_t sessionID,
+						 uint8_t interfaceVersion, MsgType type, uint8_t returnCode, const uint8_t *const data,
+						 size_t dataLen)
+{
+	const size_t headerLen = sizeof(someiphdr);
+	m_DataLen = headerLen + dataLen;
+	m_Data = new uint8_t[m_DataLen];
+	m_Protocol = SomeIP;
+	memset(m_Data, 0, headerLen);
+	memcpy(m_Data + headerLen, data, dataLen);
+
+	setPayloadLength(dataLen);
+	setServiceID(serviceID);
+	setMethodID(methodID);
+	setClientID(clientID);
+	setSessionID(sessionID);
+	setInterfaceVersion(interfaceVersion);
+	setMessageType(type);
+	setReturnCode(returnCode);
+}
+
+uint32_t SomeIpLayer::getLengthField() const
+{
+	return be32toh(getSomeIpHeader()->length);
+}
+
+void SomeIpLayer::setPayloadLength(uint32_t payloadLength)
+{
+	someiphdr *hdr = getSomeIpHeader();
+	hdr->length = htobe32(sizeof(someiphdr) - sizeof(hdr->serviceID) - sizeof(hdr->methodID) - sizeof(hdr->length) +
+						  payloadLength);
+}
+
+uint32_t SomeIpLayer::getHeaderLengthWithPayload() const
+{
+	/* Add outer header of SOME/IP to length (messageId and length field itself) */
+	return sizeof(uint32_t) * 2 + getLengthField();
+}
+
+size_t SomeIpLayer::getHeaderLen() const
+{
+	if (subProtocolFollows())
+	{
+		/* If there is a protocol coming after the normal SOME/IP header, return just the header length so PcapPlusPlus
+		handles the oncoming protocols correctly */
+		return sizeof(someiphdr);
+	}
+	else
+	{
+		/* Else, return the length of the header + payload, since it is a normal SOME/IP PDU */
+		return getHeaderLengthWithPayload();
+	}
+}
+
+bool SomeIpLayer::subProtocolFollows() const
+{
+	// TODO: handle SomeIPSD/SomeIPTP packets
+
+	return false;
+}
+
+void SomeIpLayer::computeCalculateFields()
+{
+	getSomeIpHeader()->protocolVersion = 0x01;
+
+	if (!m_NextLayer)
+		return;
+
+	// TODO: handle SomeIPSD/SomeIPTP packets

```

Comment:
I think `SomeIpLayer` doesn't need to do anything else, right?

In the next PR `SomIpSdLayer` and `SomeIpTpLayer` will have their own `computeCalculateFields()`

Reply:
only sets the protocol version, i currently moved this to the constructor which i find a better solution. In the next PR for SD/TP it will add code for setting the payload of the length field.

Reply:
We could still keep setting the version in `computeCalculateFields()`, but I don't have a strong opinion. Let's keep it as it is in the c'tor